File: 프로그래머스/2/87946. 피로도/피로도.js
let maxCnt = 0;

function solution(k, dungeons) {
    var answer = -1;
    
    const visited = new Array(dungeons.length).fill(false);
    visit(dungeons, visited, k, 0);
    
    return maxCnt;
}

function visit(dungeons, visited, k, cnt){
    maxCnt = Math.max(maxCnt, cnt);
    for (let i=0; i<dungeons.length; i++){
        if (!visited[i] && dungeons[i][0]<=k){
            visited[i]=true;
            visit(dungeons, visited, k-dungeons[i][1], cnt+1);
            visited[i]=false;
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int maxCnt = 0;
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        
        boolean[] visited = new boolean[dungeons.length];
        
        for (int i=0; i<dungeons.length; i++){
            if (k>=dungeons[i][0]){
                k-=dungeons[i][1];
                visited[i] = true;
                explore(dungeons, visited, 1, k);
                visited[i] = false;
                k+=dungeons[i][1];
            }
        }
        
        return maxCnt;
    }
    
    public void explore(int[][] dungeons, boolean[] visited, int cnt, int k){
        maxCnt = Math.max(cnt, maxCnt);
        
        for (int i=0; i<dungeons.length; i++){
            if (!visited[i]){
                if (k>=dungeons[i][0]){
                    k-=dungeons[i][1];
                    visited[i] = true;
                    explore(dungeons, visited, cnt+1, k);
                    visited[i] = false;
                    k+=dungeons[i][1];
                }
            }
        }
    }
}